src/Spice86.Core/Emulator/InterruptHandlers/Dos/DosInt21Handler.cs

```diff
@@ -1069,4 +1074,105 @@ private void SetStateFromDosFileOperationResult(bool calledFromVm, DosFileOperat
+            uint addressToEdit = MemoryUtils.ToPhysicalAddress(address.Segment, address.Offset) + physicalStartAddress;
+            Memory.UInt16[addressToEdit] += startSegment;
+        }
+    }
+
+    /// <summary>
+    /// Sets up the CPU to execute the loaded program.
+    /// </summary>
+    /// <param name="exeFile">The EXE file that was loaded.</param>
+    /// <param name="startSegment">The starting segment address of the program.</param>
+    /// <param name="pspSegment">The segment address of the program's PSP (Program Segment Prefix).</param>
+    private void SetupCpuForExe(ExeFile exeFile, ushort startSegment, ushort pspSegment) {
+        // MS-DOS uses the values in the file header to set the SP and SS registers and
+        // adjusts the initial value of the SS register by adding the start-segment
+        // address to it.
+        State.SS = (ushort)(exeFile.InitSS + startSegment);
+        State.SP = exeFile.InitSP;
+
+        // Make DS and ES point to the PSP
+        State.DS = pspSegment;
+        State.ES = pspSegment;
+
+        State.InterruptFlag = true;
+
+        // Finally, MS-DOS reads the initial CS and IP values from the program's file
+        // header, adjusts the CS register value by adding the start-segment address to
+        // it, and transfers control to the program at the adjusted address.
+        SetEntryPoint((ushort)(exeFile.InitCS + startSegment), exeFile.InitIP);
+    }
+    
+    /// <summary>
+    /// Sets the entry point of the loaded file to the specified segment and offset values.
+    /// </summary>
+    /// <param name="cs">The segment value of the entry point.</param>
+    /// <param name="ip">The offset value of the entry point.</param>
+    private void SetEntryPoint(ushort cs, ushort ip) {
+        State.CS = cs;
+        State.IP = ip;
+        if (LoggerService.IsEnabled(LogEventLevel.Verbose)) {
+            LoggerService.Verbose("Program entry point is {ProgramEntry}", ConvertUtils.ToSegmentedAddressRepresentation(cs, ip));
+        }
+    }
+    
+    private void LoadHostFileAsCom(string hostFile, string? arguments, ushort startSegment) {
+        new PspGenerator(Memory, _dos.EnvironmentVariables, _dosMemoryManager, _dosFileManager).GeneratePsp(startSegment, arguments);
+        byte[] com = File.ReadAllBytes(hostFile);
+        uint physicalStartAddress = MemoryUtils.ToPhysicalAddress(startSegment, ComOffset);
+        Memory.LoadData(physicalStartAddress, com);
+
+        // Make DS and ES point to the PSP
+        State.DS = startSegment;
+        State.ES = startSegment;
+        SetEntryPoint(startSegment, ComOffset);
+        State.InterruptFlag = true;
+    }
+
+    internal void LoadHostFile(string hostFile, string? arguments, ushort startSegment, bool isCom) {
+        if (isCom) {
+            LoadHostFileAsCom(hostFile, arguments, startSegment);
+        } else {
+            LoadHostFileAsExe(hostFile, arguments, startSegment);
+        }
```

Comment:
Not a bug fan of this construction.

The caller has to know if it's a com or exe, so it can just as well call the appropriate method itself.